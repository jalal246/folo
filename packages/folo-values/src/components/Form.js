import React from "react";

import registry from "../valuesStore";

const Form = ({
  component: FormComponent = "form",
  onSubmit: onSubmitProps,
  children,
  ...rest
}) => {
  function onSubmit(e) {
    e.preventDefault();

    if (typeof onSubmitProps === "function") {
      onSubmitProps(e, registry.getAll());
    }
  }

  return (
    <FormComponent onSubmit={onSubmit} {...rest}>
      {children}
    </FormComponent>
  );
};

export default Form;
