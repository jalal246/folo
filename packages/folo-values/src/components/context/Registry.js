class Registry {
  constructor() {
    this.btnGroup = new Set();
    this.datatObj = {};

    this.registerCellInfo = this.registerCellInfo.bind(this);
  }

  /**
   * Add cell to datatObj
   * Add groupName to btnGroup
   *
   * This function will be called when cells mount
   * after componentDidMount, the data in datatObj will be moved to state
   * datatObj is a temp object holds value to avoid update state while rendeing
   * in this case will init all cells in datatObj until rendering happens
   * then update the state so all the values update happen in state
   *
   * @param {object} cell - new cell that should be register
   * @param {string} cell.nameRef   key for value
   * @param {string||boolean} cell.initValue value
   * @param {string} cell.groupName group name in case the cell is group-toggle
   */
  registerCellInfo({ nameRef, initValue, groupName }) {
    this.datatObj[nameRef] = initValue;

    // if it has group, handle it
    if (groupName) {
      /*
       * check if group name not exist then add it and create its own set
       * then add the cell to its group
       */

      if (!this.btnGroup.has(groupName)) {
        // add it because it is new groupÿ°
        this.btnGroup.add(groupName);

        // create new set for the group
        this.btnGroup[groupName] = new Set();
      }

      if (!this.btnGroup[groupName].has(nameRef)) {
        // then add the cell name to where its belong
        // to its group
        this.btnGroup[groupName].add(nameRef);
      }
    }
  }
}

export default Registry;
