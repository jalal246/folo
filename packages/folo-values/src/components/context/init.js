import React from "react";
import { Registry } from "@folo/store/src";

function init() {
  const StoreContext = React.createContext({
    values: {},
    updateCellValue() {},
  });

  const registry = new Registry();

  const { Consumer, Provider } = StoreContext;

  return { registry, Consumer, Provider };
}

export default init;
