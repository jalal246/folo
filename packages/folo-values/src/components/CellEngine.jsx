import React, { Component } from "react";
import PropTypes from "prop-types";

import withcontext from "@folo/withcontext";

import { ValuesConsumer } from "./context";

const propTypes = {
  /**
   * value ref to the element
   * value or checked; depends on the type
   * detected by parent Cell
   */
  valueRef: PropTypes.string.isRequired,
  initValue: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,

  /**
   * only true when cell is button
   */
  isInput: PropTypes.bool.isRequired,

  /**
   * group name in case the cell is group-toggle
   * this is only valid for boolean cells
   */
  groupName: PropTypes.string,

  /**
   * key used to store value in values object
   * that owned by context
   * orginal name is valueKey
   * changing its name after being
   */
  nameRef: PropTypes.string.isRequired,

  /**
   * flag chane its value when the parent Cell render
   * this flag is created for props that changed in paren level only
   * which require this component to update
   */
  isCellUpdated: PropTypes.bool.isRequired,

  /**
   * custom render-component
   */
  CellComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func])
    .isRequired,

  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,

  /**
   * rest object contains extra props
   * passed by user to Cell
   */
  rest: PropTypes.objectOf(
    PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string,
      PropTypes.objectOf(PropTypes.string)
    ])
  ).isRequired,

  /**
   * a context function
   * trigger the context to when a updating the value
   */
  updateCellValue: PropTypes.func.isRequired,

  /**
   * cells values, owned by context
   */
  values: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
  )
};

const defaultProps = {
  groupName: null,
  values: {}
};

/**
 * manage value updates for all cell types
 * all controlled
 */
class CellEngine extends Component {
  constructor(props) {
    super(props);

    const { initValue } = props;

    this.state = { localValue: initValue };

    this.handleEvent = this.handleEvent.bind(this);
  }

  shouldComponentUpdate(nextProps, nextState) {
    const {
      nameRef,
      groupName,
      isCellUpdated,
      values: { [nameRef]: contextValue }
    } = this.props;
    const {
      values: { [nameRef]: nextContextValue },
      isCellUpdated: nextCellUpdated
    } = nextProps;

    const { localValue } = this.state;

    if (groupName && contextValue !== nextContextValue) {
      this.setState({
        localValue: nextContextValue
      });
    }
    return (
      localValue !== nextState.localValue || isCellUpdated !== nextCellUpdated
    );
  }

  handleEvent(e) {
    const {
      nameRef,
      valueRef,
      isInput,
      groupName,
      updateCellValue,
      onBlur,
      onChange
    } = this.props;
    const {
      target: { [valueRef]: newValue },
      type
    } = e;

    if (type === "blur") {
      updateCellValue({
        nameRef,
        newValue,
        groupName
      });
      onBlur(e);
    } else {
      this.setState({
        localValue: newValue
      });
      if (!isInput) {
        updateCellValue({
          nameRef,
          newValue,
          groupName
        });
      }
      onChange(e);
    }
  }

  render() {
    // console.log("CellEngine update");
    const { CellComponent, valueRef, rest, children } = this.props;

    const { localValue } = this.state;

    const { handleEvent } = this;

    return (
      <CellComponent
        {...{ [valueRef]: localValue }}
        onChange={handleEvent}
        onBlur={handleEvent}
        {...rest}
      >
        {children}
      </CellComponent>
    );
  }
}

CellEngine.propTypes = propTypes;
CellEngine.defaultProps = defaultProps;

export { CellEngine as PureCellEngine };

export default withcontext({
  Component: CellEngine,
  Consumer: ValuesConsumer,
  contextProps: ["updateCellValue", "values"]
});
