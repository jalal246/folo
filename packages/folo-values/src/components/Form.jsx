import React from "react";
import PropTypes from "prop-types";

import withContext from "@folo/withContext";

import { ValuesConsumer } from "./context";

const propTypes = {
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  onSubmit: PropTypes.func,
  getContextValues: PropTypes.func
};

const defaultProps = {
  component: "form",
  onSubmit() {},
  getContextValues() {}
};

class Form extends React.PureComponent {
  onSubmit = e => {
    e.preventDefault();
    const { getContextValues, onSubmit } = this.props;

    onSubmit(e, getContextValues());
  };

  render() {
    const {
      component: FormComponent,
      onSubmit,
      getContextValues,
      children,
      ...other
    } = this.props;
    return (
      <FormComponent onSubmit={this.onSubmit} {...other}>
        {children}
      </FormComponent>
    );
  }
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

export { Form as PureForm };

export default withContext({
  Component: Form,
  Consumer: ValuesConsumer,
  contextProps: ["getContextValues"]
});
