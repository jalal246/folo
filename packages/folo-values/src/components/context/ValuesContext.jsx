import React, { createContext } from "react";
import Registry from "./Registry";

const ValuesContext = createContext({
  values: {},
  /* istanbul ignore next */
  updateCellValue() {}
  // registerCellInfo() {},
  // getContextValues() {}
});

const { Provider, Consumer: ValuesConsumer } = ValuesContext;

class ValuesProvider extends React.Component {
  constructor(props) {
    super(props);

    this.Registry = new Registry();

    this.state = {
      values: {},
      isGroupValuesUpdate: false
    };

    this.getContextValues = this.getContextValues.bind(this);
    this.updateCellValue = this.updateCellValue.bind(this);
  }

  componentDidMount() {
    /*
     * This wont update the component
     * just set collected data obj as state
     * */
    const { datatObj, reset } = this.Registry;
    this.setState({ values: datatObj });
    // clear it
    reset();
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { isGroupValuesUpdate } = this.state;

    return isGroupValuesUpdate !== nextState.isGroupValuesUpdate;
  }

  /**
   *
   * This function will be called when trigger submit
   * which returns state.values
   *
   * @return {{value :string||boolean}}
   */
  getContextValues() {
    const { values } = this.state;
    return values;
  }

  /**
   * update cell value in the state
   *
   * @param {object} cell - new cell that should be register
   * @param {string} cell.nameRef   key for value
   * @param {string||boolean} cell.initValue value
   * @param {string} cell.groupName group name in case the cell is group-toggle
   */
  updateCellValue({ nameRef, newValue, groupName }) {
    const {
      values: { [nameRef]: oldValue }
    } = this.state;

    // dont update if it is the same value
    if (oldValue === newValue) {
      return;
    }

    this.setState(ps => {
      const newValuesHolder = {};
      newValuesHolder[nameRef] = newValue;
      let { isGroupValuesUpdate } = ps;

      if (groupName) {
        isGroupValuesUpdate = !isGroupValuesUpdate;

        if (newValue !== false) {
          // update group of values

          // toggle group values
          const { btnGroup } = this.Registry;

          btnGroup[groupName].forEach(cellNameRef => {
            // toggle all except the targeted key name which called nameRef
            // since we already changed its value above
            if (cellNameRef !== nameRef) {
              newValuesHolder[cellNameRef] = !newValue;
            }
          });
        }
      }
      return {
        values: { ...ps.values, ...newValuesHolder },
        isGroupValuesUpdate
      };
    });
  }

  render() {
    // console.log("ValuesContext update");
    const { values } = this.state;

    const { children } = this.props;

    const {
      getContextValues,
      updateCellValue,
      Registry: { registerCellInfo }
    } = this;

    return (
      <Provider
        value={{
          updateCellValue,
          registerCellInfo,
          getContextValues,
          values
        }}
      >
        {children}
      </Provider>
    );
  }
}

export { ValuesProvider, ValuesConsumer };
