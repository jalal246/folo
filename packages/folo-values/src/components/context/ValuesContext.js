import React from "react";
import { updater } from "@folo/store/src";

import init from "./init";

const { registry, Consumer: ValuesConsumer, Provider } = init();

const ValuesProvider = ({ children }) => {
  const [setState, state] = React.useState({
    values: {},
    isGroupValuesUpdate: false,
  });

  React.useEffect(() => {
    setState({ values: registry.dataObj });
  }, []);

  /**
   * update cell value in the state
   *
   * @param {object} cell - new cell that should be register
   * @param {string} cell.nameRef   key for value
   * @param {string||boolean} cell.initValue value
   * @param {string} cell.groupName group name in case the cell is group-toggle
   */
  function updateCellValue({ nameRef, newValue, groupName }) {
    const {
      values: { [nameRef]: oldValue },
    } = state;

    // don't update if it is the same value
    if (oldValue !== newValue) {
      const { btnGroup } = registry;

      setState(({ values, isGroupValuesUpdate }) =>
        updater({
          values,
          isGroupValuesUpdate,
          btnGroup,
          nameRef,
          newValue,
          groupName,
        })
      );
    }
  }

  const { values } = state;

  const value = {
    registerCellInfo: registry.registerCellInfo,
    updateCellValue,
    getContextValues: () => values,
    values,
  };

  return <Provider value={value}>{children}</Provider>;
};

export { ValuesProvider, ValuesConsumer };
