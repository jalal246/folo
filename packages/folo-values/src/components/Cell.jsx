import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import withcontext from "@folo/withcontext";
import { keyGenerator } from "@folo/utils";

import CellEngine from "./CellEngine";
import { ValuesConsumer } from "./context";

import {
  VALUE,
  CHECKED,
  TEXT,
  SELECT,
  LIST,
  CHECKBOX,
  RADIO,
  INPUT
} from "../constants";

// import componentShape from "../shapes/componentShape";
// console.log("withcontext>>>>>>>>>>>>>>>", withcontext);

const propTypes = {
  /**
   * custom render-component
   */
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

  /**
   * key used to store value in values object
   * this will be later nameRef after being processed
   */
  valueKey: PropTypes.string,

  /**
   * init value if it is string
   */
  value: PropTypes.string,

  /**
   * init value if it is boolean
   */
  checked: PropTypes.bool,

  id: PropTypes.string,
  type: PropTypes.string,

  /**
   * group name in case the cell is group-toggle
   * this is only valid for boolean cells
   */
  groupName: PropTypes.string,

  children: PropTypes.node,

  /**
   * supposed to be context function
   * helps to register values and key reference
   * beacuse context values is not aware of data we have, yet.
   */
  registerCellInfo: PropTypes.func,

  onChange: PropTypes.func,
  onBlur: PropTypes.func
};

const defaultProps = {
  component: null,
  valueKey: null,
  value: "",
  checked: false,
  id: undefined,
  type: TEXT,
  groupName: null,
  children: null,
  registerCellInfo() {},
  onChange() {},
  onBlur() {}
};

/**
 * Gets the cell type
 * returns booleans type flage.
 *
 * @param {string} type
 * @param {boolean} checked
 * @param {string} value
 * @return {{isInput:boolean, valueRef: string, initValue: string||boolean, RecommendedComponent: string }}
 */
function recognizeCellProps(type, checked, value) {
  // only true when cell is button
  let isInput = false;

  // input or select
  let RecommendedComponent = INPUT;

  // value ref to the element: value or checked; depends on the type
  let valueRef = VALUE;

  // is it boolean or string; depends on the type
  let initValue = value;

  if (type === SELECT || type === LIST) {
    RecommendedComponent = SELECT;
  } else if (type === CHECKBOX || type === RADIO) {
    valueRef = CHECKED;
    initValue = checked;
  } else {
    isInput = true;
  }
  return {
    isInput,
    valueRef,
    initValue,
    RecommendedComponent
  };
}

/**
 * mainly reposible for user props
 * handling cell type, init value and pass it to CellEngine
 * update when cell basic change
 * like attr
 */
class Cell extends PureComponent {
  constructor(props) {
    super(props);
    this.isCellUpdated = false;
  }

  render() {
    // console.log("Cell update");

    const {
      component: userComponent,
      valueKey,
      value,
      checked,
      id = keyGenerator("autoID"),
      type,
      groupName,
      children,
      registerCellInfo,
      onChange,
      onBlur,
      ...rest
    } = this.props;

    const {
      valueRef,
      isInput,
      initValue,
      RecommendedComponent
    } = recognizeCellProps(type, checked, value);

    const nameRef =
      valueKey || `${type}_${id}${groupName ? `_${groupName}` : ""}`;

    // register cell info in context state
    registerCellInfo({
      nameRef,
      initValue,
      groupName
    });

    this.isCellUpdated = !this.isCellUpdated;

    return (
      <CellEngine
        id={id}
        type={type}
        valueRef={valueRef}
        initValue={initValue}
        isInput={isInput}
        groupName={groupName}
        nameRef={nameRef}
        isCellUpdated={this.isCellUpdated}
        CellComponent={userComponent || RecommendedComponent}
        onChange={onChange}
        onBlur={onBlur}
        rest={rest}
      >
        {children}
      </CellEngine>
    );
  }
}

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export { Cell as PureCell };

export default withcontext({
  Component: Cell,
  Consumer: ValuesConsumer,
  contextProps: ["registerCellInfo"]
});
