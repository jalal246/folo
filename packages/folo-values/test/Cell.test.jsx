import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";

import { ValuesProvider } from "../src/components/context";
import Cell from "../src/components/Cell";
import Form from "../src/components/Form";

const TEST_DEFAULT = "defalut";
const TEST_BTN_1 = "btn1";
const TEST_BTN_2 = "btn2";
const TEST_FORM = "form";

const ID_1 = "unique";
const ID_2 = "unique2";

const VAL_KEY = "myRefName";
const GROUP_NAME = "gr";

const NEW_VAL = "hello!";

const onSubmitMock = jest.fn((e, value) => value);

const App = () => (
  <ValuesProvider>
    <Form data-testid={TEST_FORM} onSubmit={onSubmitMock}>
      <Cell id={ID_2} data-testid={TEST_DEFAULT} />
      <Cell />
      <Cell
        type="checkbox"
        groupName={GROUP_NAME}
        id={ID_1}
        data-testid={TEST_BTN_1}
      />
      <Cell
        type="checkbox"
        checked
        groupName={GROUP_NAME}
        valueKey={VAL_KEY}
        data-testid={TEST_BTN_2}
      />
    </Form>
  </ValuesProvider>
);

const { baseElement, getByTestId } = render(<App />);

const txt = getByTestId(TEST_DEFAULT);

const btn1 = getByTestId(TEST_BTN_1);
const btn2 = getByTestId(TEST_BTN_2);

const form = getByTestId(TEST_FORM);

describe("Cell#CellEngine & Form", () => {
  it("returns expected values and name attr", () => {
    expect(baseElement).toMatchSnapshot();
  });

  it("sets one btn to false deesnt toggle the other one which is false", () => {
    // from
    expect(btn1.checked).toBe(false);
    expect(btn2.checked).toBe(true);

    fireEvent.blur(btn2, { target: { checked: false } });

    // to
    expect(btn1.checked).toBe(false);
    expect(btn2.checked).toBe(false);
  });

  it("sets one btn to true deesnt toggle the other one which is false", () => {
    // from
    expect(btn1.checked).toBe(false);
    expect(btn2.checked).toBe(false);

    fireEvent.blur(btn2, { target: { checked: true } });

    // to
    expect(btn1.checked).toBe(false);
    expect(btn2.checked).toBe(true);
  });

  it("sets one btn to true, toggles the other to false", () => {
    // from
    expect(btn1.checked).toBe(false);
    expect(btn2.checked).toBe(true);

    fireEvent.blur(btn1, { target: { checked: true } });

    // to
    expect(btn1.checked).toBe(true);
    expect(btn2.checked).toBe(false);
  });

  it("sets same value to btn", () => {
    // from
    expect(btn1.checked).toBe(true);
    expect(btn2.checked).toBe(false);

    fireEvent.blur(btn1, { target: { checked: true } });

    // to
    expect(btn1.checked).toBe(true);
    expect(btn2.checked).toBe(false);
  });

  it("sets new value to text input", () => {
    // from
    expect(txt.value).toBe("");

    fireEvent.change(txt, { target: { value: NEW_VAL } });

    // to
    expect(txt.value).toBe(NEW_VAL);
  });

  it("sets same value to text input", () => {
    // from
    expect(txt.value).toBe(NEW_VAL);

    fireEvent.change(txt, { target: { value: NEW_VAL } });

    // to
    expect(txt.value).toBe(NEW_VAL);
  });

  it("return context value when sbmit", () => {
    fireEvent.submit(form, {});

    // to
    expect(onSubmitMock).toHaveReturnedWith(
      expect.objectContaining({
        text_unique2: "",
        // unknown_valueKey_created_at_1547750212936: "",
        checkbox_unique_gr: true,
        myRefName: false
      })
    );

    cleanup();
  });
});
