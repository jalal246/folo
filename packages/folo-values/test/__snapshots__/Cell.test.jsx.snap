// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell#CellEngine & Form returns expected values and name attr 1`] = `
<body>
  <div>
    <form
      data-testid="form"
    >
      <input
        data-testid="defalut"
        id="unique2"
        type="text"
        value=""
      />
      <input
        type="text"
        value=""
      />
      <input
        data-testid="btn1"
        id="unique"
        type="checkbox"
      />
      <input
        checked=""
        data-testid="btn2"
        type="checkbox"
      />
    </form>
  </div>
</body>
`;
